File: snippets/javascript/knowledge_base.js
/**
 * Perfex CRM REST API — Knowledge Base examples (Node.js / fetch)
 * Module: https://themesic.com/product/rest-api-module-for-perfex-crm-connect-your-perfex-crm-with-third-party-applications/
 *
 * Node 18+ has global fetch & FormData. No dependencies required.
 */

const BASE = "https://yourdomain.com/api";
const TOKEN = "YOUR_API_TOKEN";

async function listArticles() {
  const res = await fetch(`${BASE}/knowledge_base`, {
    headers: { authtoken: TOKEN },
  });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return res.json();
}

// Filter articles by group.
async function listArticlesByGroup(groupId = 1) {
  const res = await fetch(`${BASE}/knowledge_base?group_id=${groupId}`, {
    headers: { authtoken: TOKEN },
  });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return res.json();
}

async function getArticle(id = 12) {
  const res = await fetch(`${BASE}/knowledge_base/${id}`, {
    headers: { authtoken: TOKEN },
  });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return res.json();
}

async function createArticle() {
  // POST/PUT send multipart/form-data
  const form = new FormData();
  form.append("subject", "How to reset your password");
  form.append("description", "Open Settings > Security and click Reset password.");
  form.append("articlegroup", "1");
  form.append("active", "1");

  const res = await fetch(`${BASE}/knowledge_base`, {
    method: "POST",
    headers: { authtoken: TOKEN },
    body: form,
  });
  return { status: res.status, data: await res.json() };
}

async function updateArticle(id = 12) {
  const form = new FormData();
  form.append("subject", "How to reset your password (updated)");

  const res = await fetch(`${BASE}/knowledge_base/${id}`, {
    method: "PUT",
    headers: { authtoken: TOKEN },
    body: form,
  });
  return { status: res.status, data: await res.json() };
}

async function deleteArticle(id = 12) {
  const res = await fetch(`${BASE}/knowledge_base/${id}`, {
    method: "DELETE",
    headers: { authtoken: TOKEN },
  });
  return { status: res.status, data: await res.json() };
}

async function listGroups() {
  const res = await fetch(`${BASE}/knowledge_base/groups`, {
    headers: { authtoken: TOKEN },
  });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return res.json();
}

async function createGroup() {
  const form = new FormData();
  form.append("name", "Billing");

  const res = await fetch(`${BASE}/knowledge_base/groups`, {
    method: "POST",
    headers: { authtoken: TOKEN },
    body: form,
  });
  return { status: res.status, data: await res.json() };
}

(async () => {
  console.log(await listArticles());
  console.log(await listArticlesByGroup(1));
  console.log(await getArticle(12));
  console.log(await createArticle());
  console.log(await updateArticle(12));
  console.log(await deleteArticle(12));
  console.log(await listGroups());
  console.log(await createGroup());
})();

File: snippets/python/knowledge_base.py
"""
Perfex CRM REST API — Knowledge Base examples (Python / requests)
Module: https://themesic.com/product/rest-api-module-for-perfex-crm-connect-your-perfex-crm-with-third-party-applications/

Install dependency:  pip install requests
"""
import requests

BASE = "https://yourdomain.com/api"
TOKEN = "YOUR_API_TOKEN"
HEADERS = {"authtoken": TOKEN}


def list_articles():
    r = requests.get(f"{BASE}/knowledge_base", headers=HEADERS, timeout=30)
    r.raise_for_status()
    return r.json()


def list_articles_in_group(group_id=1):
    # Filter articles by their group.
    params = {"group_id": group_id}
    r = requests.get(f"{BASE}/knowledge_base", headers=HEADERS, params=params, timeout=30)
    r.raise_for_status()
    return r.json()


def get_article(article_id=12):
    r = requests.get(f"{BASE}/knowledge_base/{article_id}", headers=HEADERS, timeout=30)
    r.raise_for_status()
    return r.json()


def create_article():
    # POST/PUT send multipart/form-data — pass fields via `data=`
    payload = {
        "subject": "How to reset your password",
        "description": "<p>Open Settings &gt; Security and click Reset.</p>",
        "articlegroup": 1,
        "active": 1,
    }
    r = requests.post(f"{BASE}/knowledge_base", headers=HEADERS, data=payload, timeout=30)
    return r.status_code, r.json()


def update_article(article_id=12):
    payload = {"active": 0}
    r = requests.put(f"{BASE}/knowledge_base/{article_id}", headers=HEADERS, data=payload, timeout=30)
    return r.status_code, r.json()


def delete_article(article_id=12):
    r = requests.delete(f"{BASE}/knowledge_base/{article_id}", headers=HEADERS, timeout=30)
    return r.status_code, r.json()


def list_groups():
    r = requests.get(f"{BASE}/knowledge_base/groups", headers=HEADERS, timeout=30)
    r.raise_for_status()
    return r.json()


def create_group():
    payload = {"name": "Billing"}
    r = requests.post(f"{BASE}/knowledge_base/groups", headers=HEADERS, data=payload, timeout=30)
    return r.status_code, r.json()


if __name__ == "__main__":
    print(list_articles())
    print(list_articles_in_group(1))
    print(get_article(12))
    print(create_article())
    print(update_article(12))
    print(list_groups())
    print(create_group())
    print(delete_article(12))
